Realm/RLMQueryUtil.mm

```diff
@@ -980,6 +989,34 @@ void update_query_with_column_expression(RLMSchema *schema, RLMObjectSchema *des
                             std::move(left), std::move(right));
 }
 
+NSExpression *simplify_self_value_for_key_path_function_expression(NSExpression *expression) {
+    // -[NSPredicate predicateWithSubtitutionVariables:] results in function expressions of the form [SELF valueForKeyPath:]
+    // that update_query_with_predicate cannot handle. Replace such expressions with equivalent NSKeyPathExpressionType expressions.
+    if (expression.expressionType == NSFunctionExpressionType
+        && expression.operand.expressionType == NSEvaluatedObjectExpressionType
+        && [expression.function isEqualToString:@"valueForKeyPath:"]) {
+        if (NSString *keyPath = expression.arguments.firstObject.keyPath) {
+            return [NSExpression expressionForKeyPath:keyPath];
+        }
+    }
+    return expression;
+}
+
+void update_query_with_subquery_expression(RLMSchema *schema, RLMObjectSchema *objectSchema, realm::Query& query,
+                                           NSExpression *subqueryExpression, NSPredicateOperatorType operatorType, NSUInteger value) {
```

Comment:
Why is value NSUInteger? The value passed to it is signed and it's cast back to int64_t when it's used.